articles/unique_network_flows.md

```diff
@@ -0,0 +1,171 @@
+
+N1 contains publishers P1 and P2. P1 publishes video data whereas P2 publishes battery status data.
+
+N2 contains subscribers S1 and S2. S1 receives video from P1 and performs real-time object detection. S2 receives battery data from P2 and performs non-real-time battery management.
+
+The link P1-S1 requires low-latency QoS from the 5G network, say a maximum delay of 5ms. P2-S2 requires default latency QoS i.e., 300ms. Then, by construction, since the flow identifiers of P1-S1 and P2-S2 links are the same, they cannot be assigned the required QoS by the network. Both P1-S1 and P2-S2 can either be assigned QoS with 5ms delay or with 300ms delay. The former case represents a waste of network resources, the latter case degrades performance.
+
+## Proposed Solution
+
+Our proposal to solve the problem is to make the flow identifiers of publishers and subscribers in communicating nodes unique using existing IP header fields.
+
+We believe there are at least three architectural options to implement unique network flows. These are shown in the figure below.
+
+![Architecture options for unique network flows](./unique-network-flows-arch-options.png)
+
+### Option 1: Unique flow ID from publisher/subscriber
+
+We construct a new QoS policy called *Network Flow* as a support structure to enable unique identification of flows.
+
+The Network Flow QoS policy is parameterized by just one parameter -- a 32-bit unsigned integer called `flow_id`. The default `flow_id` is zero.
+
+Each publisher or subscriber that requires a specific QoS from the network is required to set the `flow_id` parameter to value unique within the node. The `flow_id` can be computed internally in the node using simple UID generation schemes (atomic integer counters or random integer sampling without duplicates) or supplied by an external component.
+
+The example C++ snippet below shows a node creating a publisher (`pub1`)  and a subscriber (`sub1`) with unique `flow_id` values. Subscriber `sub2` is created with the default Network Flow QoS policy (`flow_id` = 0).
+
+```cpp
+auto pub1_qos = rclcpp::QoS(rclcpp::KeepLast(10)).network_flow(0xf4688c52);
+pub1 = create_publisher<std_msgs::msg::UInt64>("pub1_topic", pub1_qos);
+
+auto sub1_qos = rclcpp::QoS(rclcpp::KeepLast(1)).network_flow(0x88181c45);
+sub1 = create_subscription<std_msgs::msg::String>("sub1_topic", sub1_qos, std::bind(&new_sub1_message, this, _1));
+
+auto sub2_qos = rclcpp::QoS(rclcpp::KeepLast(1));
+sub2 = create_subscription<std_msgs::msg::String>("sub2_topic", sub2_qos, std::bind(&new_sub2_message, this, _1));
+```
+
+The RMW implementation has several options to convert `flow_id` to a unique flow identifier visible in packet headers. We list three candidate options.
+
+If the node is communicating using IPv6, then the lower 20-bits of `flow_id` can be transferred to the Flow Label field. This creates a unique 3-tuple.
```

Comment:
what if the lower 20-bits are the same? e.g) in the same node, id-A is 0x0badcafe id-B is 0x0bbdcafe. in this case lower 20-bits are the exactly same.

Reply:
Sharp observation! It would be problematic if the lower 20-bits are the same across two or more unique flow IDs within a node. We expect that the RMW will document how it creates the unique flow ID in the IP header and that the application programmer generates unique flow IDs accordingly. For example, if the RMW documentation says that the unique flow ID is copied over to an N-bit field in the IP header, then the application programmer should generate flow IDs that are N-bit unique. Other suggestions to ensure that unique flow IDs do not overlap are welcome.